metadata-provider/src/contract.rs

```diff
@@ -0,0 +1,271 @@
+        HandleMsg::SetMetadata {
+            token_id,
+            public_metadata,
+            private_metadata,
+            ..
+        } => handle_set_metadata(
+            deps,
+            env,
+            &config,
+            &token_id,
+            public_metadata,
+            private_metadata,
+        ),
+        HandleMsg::SetViewingKey { key, .. } => handle_set_key(deps, env, &config, key),
+        HandleMsg::ChangeAdmin { address, .. } => {
+            handle_change_admin(deps, env, &mut config, &address)
+        }
+    };
+    pad_handle_result(response, BLOCK_SIZE)
+}
+
+pub fn handle_set_metadata<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    _env: Env,
+    _config: &Config,
+    token_id: &str,
+    public_metadata: Option<Metadata>,
+    private_metadata: Option<Metadata>,
+) -> HandleResult {
+    if let Some(public) = public_metadata {
+        set_metadata_impl(&mut deps.storage, token_id, PREFIX_PUB_META, &public)?;
+    }
+    if let Some(private) = private_metadata {
+        set_metadata_impl(&mut deps.storage, token_id, PREFIX_PRIV_META, &private)?;
+    }
+
+    Ok(HandleResponse {
+        messages: vec![],
+        log: vec![],
+        data: Some(to_binary(&HandleAnswer::SetMetadata { status: Success })?),
+    })
+}
+
+/// Returns StdResult<()>
+///
+/// sets new metadata
+///
+/// # Arguments
+///
+/// * `storage` - a mutable reference to the contract's storage
+/// * `token` - a reference to the token whose metadata should be updated
+/// * `idx` - the token identifier index
+/// * `prefix` - storage prefix for the type of metadata being updated
+/// * `metadata` - a reference to the new metadata
+#[allow(clippy::too_many_arguments)]
+fn set_metadata_impl<S: Storage>(
+    storage: &mut S,
+    token_id: &str,
+    prefix: &[u8],
+    metadata: &Metadata,
+) -> StdResult<()> {
+    enforce_metadata_field_exclusion(metadata)?;
+    let mut meta_store = PrefixedStorage::new(prefix, storage);
+    save(&mut meta_store, &token_id.as_bytes(), metadata)?;
+    Ok(())
+}
+
+/// Returns StdResult<()>
+///
+/// makes sure that Metadata does not have both `token_uri` and `extension`
+///
+/// # Arguments
+///
+/// * `metadata` - a reference to Metadata
+fn enforce_metadata_field_exclusion(metadata: &Metadata) -> StdResult<()> {
+    if metadata.token_uri.is_some() && metadata.extension.is_some() {
+        return Err(StdError::generic_err(
+            "Metadata can not have BOTH token_uri AND extension",
+        ));
+    }
+    Ok(())
+}
+
+pub fn handle_set_key<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    _config: &Config,
+    key: String,
+) -> HandleResult {
+    let vk = ViewingKey(key.clone());
+    let message_sender = deps.api.canonical_address(&env.message.sender)?;
+    let mut key_store = PrefixedStorage::new(PREFIX_VIEW_KEY, &mut deps.storage);
+    save(&mut key_store, message_sender.as_slice(), &vk.to_hashed())?;
```

Comment:
Are you expecting a user/a frontend to interact with the metadata contract directly? I don't think that's necessarily the best idea. 

The viewing key should in my opinion be for the nft contract to query the data not for the user